coderd/keyrotate/rotate.go

```diff
@@ -0,0 +1,287 @@
+					}
+					k.logger.Debug(ctx, "deleted key",
+						slog.F("key", key.Sequence),
+						slog.F("feature", key.Feature),
+					)
+					modifiedKeys = append(modifiedKeys, deletedKey)
+				case shouldRotateKey(key, k.keyDuration, now):
+					rotatedKeys, err := k.rotateKey(ctx, tx, key)
+					if err != nil {
+						return xerrors.Errorf("rotate key: %w", err)
+					}
+					k.logger.Debug(ctx, "rotated key",
+						slog.F("key", key.Sequence),
+						slog.F("feature", key.Feature),
+					)
+					validKeys++
+					modifiedKeys = append(modifiedKeys, rotatedKeys...)
+				default:
+					// We only consider keys without a populated deletes_at field as valid.
+					// This is because under normal circumstances the deletes_at field
+					// is set during rotation (meaning a new key was generated)
+					// but it's possible if the database was manually altered to
+					// delete the new key we may be in a situation where there
+					// isn't a key to replace the one scheduled for deletion.
+					if !key.DeletesAt.Valid {
+						validKeys++
+					}
+				}
+			}
+			if validKeys == 0 {
+				k.logger.Info(ctx, "no valid keys detected, inserting new key",
+					slog.F("feature", feature),
+				)
+				latestKey, err := tx.GetLatestCryptoKeyByFeature(ctx, feature)
+				if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+					return xerrors.Errorf("get latest key: %w", err)
+				}
+
+				newKey, err := k.insertNewKey(ctx, tx, feature, now, latestKey.Sequence+1)
+				if err != nil {
+					return xerrors.Errorf("insert new key: %w", err)
+				}
+				modifiedKeys = append(modifiedKeys, newKey)
+			}
+		}
+		return nil
+	})
+}
+
+func (k *Rotator) insertNewKey(ctx context.Context, tx database.Store, feature database.CryptoKeyFeature, startsAt time.Time, sequence int32) (database.CryptoKey, error) {
+	secret, err := generateNewSecret(feature)
+	if err != nil {
+		return database.CryptoKey{}, xerrors.Errorf("generate new secret: %w", err)
+	}
+
+	newKey, err := tx.InsertCryptoKey(ctx, database.InsertCryptoKeyParams{
+		Feature:  feature,
+		Sequence: sequence,
+		Secret: sql.NullString{
+			String: secret,
+			Valid:  true,
+		},
+		// Set by dbcrypt if it's required.
+		SecretKeyID: sql.NullString{},
+		StartsAt:    startsAt.UTC(),
+	})
+	if err != nil {
+		return database.CryptoKey{}, xerrors.Errorf("inserting new key: %w", err)
+	}
+
+	k.logger.Info(ctx, "inserted new key for feature", slog.F("feature", feature))
+	return newKey, nil
+}
+
+func (k *Rotator) rotateKey(ctx context.Context, tx database.Store, key database.CryptoKey) ([]database.CryptoKey, error) {
+	// The starts at of the new key is the expiration of the old key. We set the deletes_at of the old key to a little over
+	// an hour after its set to expire so there should plenty
+	// of time for the new key to enter rotation.
+	newStartsAt := key.ExpiresAt(k.keyDuration)
+
+	newKey, err := k.insertNewKey(ctx, tx, key.Feature, newStartsAt, key.Sequence+1)
```

Comment:
I don't think its safe to assume that a rotatable key is _always_ the highest sequence number.

Reply:
The query itself could set the sequence number to the max + 1 for the feature to make this better

Reply:
I'd have to think carefully whether that's safe with RepeatableRead isolation

Reply:
Pretty sure that result in a double key insertion